### ISSUE TITLE
Potential bug in trainer.py

### ISSUE THREAD

**Issue by cjliux** [zliucr/coach#2]
I tried to deploy `coach` on my server without any changes, and it turns out that something is wrong in `train.py` when I tried to reproduce the reported results by running `python slu_main.py --exp_name coach_lstmenc --exp_id atp_50 --bidirection --freeze_emb --tgt_dm PlayMusic --n_samples 50`. 
![批注 2020-06-16 103640](https://user-images.githubusercontent.com/14961563/84727344-636f2a00-afc1-11ea-8ffb-7f6490e685a2.png)
By careful debugging, I find that the loss update procedure in the function `train_step` is somewhat problematic. The loss forwarding of three constituents of the model are done with the same model snapshot, while their gradient computation are done separately.
Somehow the code works now with some unknown reasons, maybe because I have added several lines of comments : ). But I am still very entangled about the error prompt that happened before.

**Comment by zliucr** [comment]
Hi,
Which PyTorch version did you use? This code is runnable under python 3.6 and PyTorch 1.4.0. 
I think it is okay to separate the gradient computation (however, it could be slower in the training). It would be great if you merge the loss and do gradient computation only once.

**Comment by cjliux** [comment]
I have tried merging them and updating at the end for once, though it is frustrating with F1 being around `55.1587`, far below the reported `62.41` in Table 1. As I see it, the reported results are more `Binary Slot-F1` rather than `Final Slot-F1`. However, in the unchanged version, the reproduced `Final-F1` is around `59.7138`. Therefore, I can conclude that the two implementations are different.
So, why they are different? When updated separately, the gradients we used are not the same thing.  Take the gradient of LSTM encoder w.r.t. the loss over the `slotname_predictor` for example. When updating for only once, the gradient is `\frac {\partial {old loss}}{\partial {old LSTM}}`; while updating separately, the gradient is `\frac {\partial {old loss}}{\partial {new LSTM}}`.
